verifier-sdk/src/light_transaction.rs

```diff
@@ -526,14 +518,95 @@ impl<
     }
 
     /// Calls the Merkle tree program via cpi to insert transaction leaves.
-    pub fn insert_leaves(&mut self) -> Result<()> {
+    pub fn insert_leaves<'b, 'c, 'd, 'info2, A>(
+        &mut self,
+        ctx: &Context<'b, 'c, 'd, 'info2, A>,
+    ) -> Result<()>
+    where
+        A: LightAccounts<'info2> + anchor_lang::Bumps,
+    {
         if !self.verified_proof {
             msg!("Tried to insert leaves without verifying the proof.");
             return err!(VerifierSdkError::ProofNotVerified);
         }
 
-        let transaction_merkle_tree = self.get_transaction_merkle_tree()?;
+        // Index of a new Transaction Merkle Tree in remaining accounts.
+        let index = NR_NULLIFIERS + self.input.leaves.len();
 
+        // Determine a Transaction Merkle Tree which should be used for the current
+        // transaction. It might be either:
+        //
+        // * Transaction Merkle Tree provided in the context.
+        // * A new Transaction Merkle Tree provided as a remaining account, which
+        //   usually is the case when we reached the switch threshold (255_000).
+        //   We pass a new Transaction Merkle Tree as remaining account, because
+        //   Anchor does not support passing two accounts of the same type in the
+        //   same instruction. We need a second Transacton Merkle Tree account for
+        //   the UTXOs we want to spend in the transaction are in the old
+        //   Transaction Merkle Tree. Since the old Merkle Tree is almost full we
+        //   need to insert the new utxo commitments into the new Transaction
+        //   Merkle Tree.
+        match ctx.remaining_accounts.get(index) {
+            Some(transaction_merkle_tree) => {
+                let transaction_merkle_tree = transaction_merkle_tree.to_account_info();
+                // NOTE(vadorovsky): Unfortunately, I didn't find any way to
+                // create a temporary `AccountLoader`, just for accessing the
+                // `merkle_tree_nr` field, without annoying a borrow checker,
+                // complaining that `transacion_merkle_tree` "does not live
+                // long enough" (yeah, we are absolutely fine with that).
+                //
+                // So far, hacky way of accessing it, which is basically what
+                // `AccountLoader::load()` in Anchor does, but without lifetime
+                // constraints in a separate function, is the sanest way which
+                // works.
+                let index = {
+                    let data = transaction_merkle_tree.try_borrow_data()?;
+                    if data.len() > TransactionMerkleTree::discriminator().len() {
```

Comment:
shouldn't it be as follows?
```suggestion
                    if data.len() < TransactionMerkleTree::discriminator().len() {
```